lib/config/applyrule.hpp

```diff
@@ -15,11 +16,42 @@ namespace icinga
 /**
  * @ingroup config
  */
-class ApplyRule
+class ApplyRule : public SharedObject
 {
 public:
+	DECLARE_PTR_TYPEDEFS(ApplyRule);
+
+	struct PerHost
+	{
+		std::vector<ApplyRule::Ptr> ForHost;
+		std::unordered_map<String /* service */, std::vector<ApplyRule::Ptr>> ForServices;
+	};
+
+	struct PerTypes
+	{
+		std::vector<ApplyRule::Ptr> Regular;
+		std::unordered_map<String /* host */, PerHost> Targeted;
+	};
+
+	/*
+	 * m_Rules[T::TypeInstance.get()][C::TypeInstance.get()].Regular
+	 * contains all apply rules like apply T "x" to C { ... }
+	 * which don't target only specific checkables, i.e. not just
+	 * assign where host.name == "H" || host.name == "h".
+	 *
+	 * m_Rules[T::TypeInstance.get()][Host::TypeInstance.get()].Targeted["H"].ForHost
+	 * contains all apply rules like apply T "x" to Host { ... }
+	 * which target only specific hosts incl. "H", e. g. via
+	 * assign where host.name == "H" || host.name == "h".
+	 *
+	 * m_Rules[T::TypeInstance.get()][Service::TypeInstance.get()].Targeted["H"].ForServices["S"]
+	 * contains all apply rules like apply T "x" to Service { ... }
+	 * which target only specific services incl. "H!S", e. g. via
+	 * assign where host.name == "H" && service.name == "S".
```

Comment:
> incl. "H!S"

What's this supposed to say? That the host name also has to be matched and just `service.name == "S"` won't be optimized? I find this confusing because you can't use "H!S" names anywhere so that it would be optimized by this PR.

Reply:
Doesn’t an Icinga dev know that name schema?

Doesn’t the last line clarify what it stands for?

Reply:
Yes, the syntax is known, but how it's currently written, you might try to read is something like "including when written as"  which doesn't apply here. I'd say "target only specific services on specific hosts" instead.